### PR TITLE
Changes to Mutect2 to support Mutect3

### PR DESCRIPTION
**Pull request by davidbenjamin**
@takutosato Are you able to look at this?  The two big changes here are 1) data for the M3 normal artifact filter and 2) outputting tensors for M3 as plain text rather than VCF.

### CODE REVIEW

**Comment by takutosato** [comment]
@davidbenjamin sorry I may not be able to get to it until later this week, is that acceptable?

**Comment by davidbenjamin** [comment]
@takutosato Sure, not a problem!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        }
+
+        // haplotype equivalence counts, haplotype complexity, haplotype dominance
+        final Triple<int[], int[], double[]> assemblyComplexity = AssemblyComplexity.annotate(vc, logFragmentLikelihoods);
+
+        // TODO: for now we don't really need normal reads
+        final List<List<List<Integer>>> normalReadVectorsByAllele =  FeaturizedReadSets.getReadVectors(vc, normalSamples, likelihoods, logFragmentLikelihoods, maxRefCount, maxAltCount);
+        final List<List<List<Integer>>> tumorReadVectorsByAllele =  FeaturizedReadSets.getReadVectors(vc, tumorSamples, likelihoods, logFragmentLikelihoods, maxRefCount, maxAltCount, altDownsampleMap);
+
+        // ref reads have already been downsampled by the read featurizer
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java:179]: both ref and alt reads have been downsampled by the read featurizer?

**Review by takutosato — APPROVED** [review]
Sorry I took so long. Just minor comments and a couple self-serving variable rename requests

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2IntegrationTest.java [hunk 2]
```diff
@@ -301,26 +304,16 @@ public void testTumorOnly() {
...
-    // run tumor-only using our mini gnomAD on NA12878, which is not a tumor
     @Test
-    public void testTrainingDataMode() {
+    public void testMutect3Dataset() {
         Utils.resetRandomGenerator();
         final File tumor = new File(NA12878_20_21_WGS_bam);
         final File unfilteredVcf = createTempFile("unfiltered", ".vcf");
+        final File mutect3Dataset = createTempFile("mutect3", ".data");
         
-        final List<String> assemblyComplexityKeys = new AssemblyComplexity().getKeyNames();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2IntegrationTest.java:311]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/GetNormalArtifactDataIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                .addInput(tumor)
+                .addInput(normal)
+                .addOutput(output)
+                .addReference(b37Reference)
+                .addInterval("20")
+                .add(M2ArgumentCollection.NORMAL_SAMPLE_LONG_NAME, normalSample);
+
+        runCommandLine(args);
+
+        final List<NormalArtifactRecord> records = NormalArtifactRecord.readFromFile(output);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/GetNormalArtifactDataIntegrationTest.java:46]: used it

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/GetNormalArtifactDataIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
+package org.broadinstitute.hellbender.tools.walkers.mutect;
+
+import htsjdk.variant.variantcontext.VariantContext;
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/GetNormalArtifactDataIntegrationTest.java:3]: removed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/NormalArtifactRecord.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        this.normalAltCount = normalAltCount;
+        this.normalDepth = normalDepth;
+        this.tumorAltCount = tumorAltCount;
+        this.tumorDepth = tumorDepth;
+        this.downsampling = downsampling;
+        this.type = type;
+    }
+
+
+    //----- The following two public static methods read and write contamination files
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/NormalArtifactRecord.java:59]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/NormalArtifactRecord.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    public static List<NormalArtifactRecord> readFromFile(final File tableFile) {
+        try( NormalArtifactReader reader = new NormalArtifactReader(IOUtils.fileToPath(tableFile)) ) {
+            return reader.toList();
+        } catch (IOException e){
+            throw new UserException(String.format("Encountered an IO exception while reading from %s.", tableFile));
+        }
+    }
+
+    //-------- The following methods are boilerplate for reading and writing contamination tables
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/NormalArtifactRecord.java:76]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+            throw new UserException.BadInput("Could not create dataset file writer");
+        }
+
+        this.normalSamples = Utils.nonNull(normalSamples);
+        this.trainingMode = trainingMode;
+        this.nonArtifactPerArtifact = nonArtifactPerArtifact;
+        this.maxRefCount = maxRefCount;
+        this.maxAltCount = maxAltCount;
+
+        unmatchedCounts = new EnumMap<VariantType, ArrayBlockingQueue<Integer>>(VariantType.class);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java:97]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    private final int nonArtifactPerArtifact;
+
+    // are we generating dataset for training a model or for filtering calls with a pre-trained model?
+    private final boolean trainingMode;
+
+    private final Set<String> normalSamples;
+
+    // simple method to balance data: for each k-alt-read artifact there are
+    // nonArtifactPerArtifact (downsampled) k-alt-read non-artifacts.
+    private final EnumMap<VariantType, ArrayBlockingQueue<Integer>> unmatchedCounts;
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java:81]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java:179]: thanks, fixed

**Pull request merged**